### PR TITLE
bpobj_enqueue_subobj() should copy small subobj's

### PR DESCRIPTION
**Pull request by ahrens**
### Motivation and Context

When we delete a snapshot, we consolidate some bpobj's together because
we no longer need to keep their entries in separate buckets.  This is
done in constant time by including the "sub" bpobj by reference in the
parent bpobj.

After many snapshots have been deleted, we may have many sub-bpobj's.
Usually, most sub-bpobj's don't contain many BP's.  Compared to this
small payload, the sub-bpobj is relatively heavyweight since it is a
object in the MOS.  A common scenerio on a long-lived pool is for the
vast majority of MOS objects to be small sub-bpobj's.
### Description

To improve this situation, when consolidating bpobj's together,
bpobj_enqueue_subobj() can copy the contents of small bpobj's into the
parent, and then delete the enqueued bpobj, rather than including it by
reference.  Since this copying is limited in size (to one block), the
consolidation is still constant time, though with a larger constant due
to reading in the one block of the enqueued bpobj.

This idea and mechanism are similar to how we handle "sub-subobj's".
When including a sub-bpobj by reference, if the sub-bpobj itself has
less than a block of sub-sub-bpobj's, the list of sub-sub-bpobj's is
copied to the parent bpobj's list of sub-bpobj's.

Additionally, I added a comment to describe the 3 cases:
```
/*
 * Logically add subobj's contents to the parent bpobj.
 *
 * In the most general case, this is accomplished in constant time by adding
 * a reference to subobj.  This case is used when enqueuing a large subobj:
 * +--------------+                        +--------------+
 * | bpobj        |----------------------->| subobj list  |
 * +----+----+----+----+----+              +-----+-----+--+--+
 * | bp | bp | bp | bp | bp |              | obj | obj | obj |
 * +----+----+----+----+----+              +-----+-----+-----+
 *
 * +--------------+                        +--------------+
 * | sub-bpobj    |----------------------> | subsubobj    |
 * +----+----+----+----+---------+----+    +-----+-----+--+--------+-----+
 * | bp | bp | bp | bp |   ...   | bp |    | obj | obj |    ...    | obj |
 * +----+----+----+----+---------+----+    +-----+-----+-----------+-----+
 *
 * Result: sub-bpobj added to parent's subobj list.
 * +--------------+                        +--------------+
 * | bpobj        |----------------------->| subobj list  |
 * +----+----+----+----+----+              +-----+-----+--+--+-----+
 * | bp | bp | bp | bp | bp |              | obj | obj | obj | OBJ |
 * +----+----+----+----+----+              +-----+-----+-----+--|--+
 *                                                              |
 *       /-----------------------------------------------------/
 *       v
 * +--------------+                        +--------------+
 * | sub-bpobj    |----------------------> | subsubobj    |
 * +----+----+----+----+---------+----+    +-----+-----+--+--------+-----+
 * | bp | bp | bp | bp |   ...   | bp |    | obj | obj |    ...    | obj |
 * +----+----+----+----+---------+----+    +-----+-----+-----------+-----+
 *
 *
 * In a common case, the subobj is small: its bp's and its list of subobj's
 * are each stored in a single block.  In this case we copy the subobj's
 * contents to the parent:
 * +--------------+                        +--------------+
 * | bpobj        |----------------------->| subobj list  |
 * +----+----+----+----+----+              +-----+-----+--+--+
 * | bp | bp | bp | bp | bp |              | obj | obj | obj |
 * +----+----+----+----+----+              +-----+-----+-----+
 *                          ^                                ^
 * +--------------+         |              +--------------+  |
 * | sub-bpobj    |---------^------------> | subsubobj    |  ^
 * +----+----+----+         |              +-----+-----+--+  |
 * | BP | BP |-->-->-->-->-/               | OBJ | OBJ |-->-/
 * +----+----+                             +-----+-----+
 *
 * Result: subobj destroyed, contents copied to parent:
 * +--------------+                        +--------------+
 * | bpobj        |----------------------->| subobj list  |
 * +----+----+----+----+----+----+----+    +-----+-----+--+--+-----+-----+
 * | bp | bp | bp | bp | bp | BP | BP |    | obj | obj | obj | OBJ | OBJ |
 * +----+----+----+----+----+----+----+    +-----+-----+-----+-----+-----+
 *
 *
 * If the subobj has many BP's but few subobj's, we can copy the sub-subobj's
 * but retain the sub-bpobj:
 * +--------------+                        +--------------+
 * | bpobj        |----------------------->| subobj list  |
 * +----+----+----+----+----+              +-----+-----+--+--+
 * | bp | bp | bp | bp | bp |              | obj | obj | obj |
 * +----+----+----+----+----+              +-----+-----+-----+
 *                                                           ^
 * +--------------+                        +--------------+  |
 * | sub-bpobj    |----------------------> | subsubobj    |  ^
 * +----+----+----+----+---------+----+    +-----+-----+--+  |
 * | bp | bp | bp | bp |   ...   | bp |    | OBJ | OBJ |-->-/
 * +----+----+----+----+---------+----+    +-----+-----+
 *
 * Result: sub-sub-bpobjs and subobj added to parent's subobj list.
 * +--------------+                     +--------------+
 * | bpobj        |-------------------->| subobj list  |
 * +----+----+----+----+----+           +-----+-----+--+--+-----+-----+-----+
 * | bp | bp | bp | bp | bp |           | obj | obj | obj | OBJ | OBJ | OBJ |
 * +----+----+----+----+----+           +-----+-----+-----+-----+-----+--|--+
 *                                                                       |
 *       /--------------------------------------------------------------/
 *       v
 * +--------------+
 * | sub-bpobj    |
 * +----+----+----+----+---------+----+
 * | bp | bp | bp | bp |   ...   | bp |
 * +----+----+----+----+---------+----+
 */
```

### How Has This Been Tested?

some manual testing

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/bpobj.c [hunk 1]
```diff
@@ -406,61 +493,110 @@ bpobj_enqueue_subobj(bpobj_t *bpo, uint64_t subobj, dmu_tx_t *tx)
...
+		dmu_buf_t *subdb;
+		uint64_t numsubsub = subbpo.bpo_phys->bpo_num_subobjs;
+
+		VERIFY0(dmu_buf_hold(bpo->bpo_os, subsubobjs,
+		    0, FTAG, &subdb, 0));
+		/*
+		 * Make sure that we are not asking dmu_write()
+		 * to write more data than we have in our buffer.
+		 */
+		VERIFY3U(subdb->db_size, >=,
```
[the remark this answers is not in the document]
Comment (ahrens) [reply module/zfs/bpobj.c]: That is not guaranteed; the dmu_write() can straddle a block boundary and is handled by dmu_write().

File: module/zfs/bpobj.c [hunk 2]
```diff
@@ -406,61 +493,110 @@ bpobj_enqueue_subobj(bpobj_t *bpo, uint64_t subobj, dmu_tx_t *tx)
...
+		ASSERT(copy_subsub);
+		VERIFY0(dmu_buf_hold(bpo->bpo_os, subobj,
+		    0, FTAG, &bps, 0));
+
+		/*
+		 * Make sure that we are not asking dmu_write()
+		 * to write more data than we have in our buffer.
+		 */
+		VERIFY3U(bps->db_size, >=, numbps * sizeof (blkptr_t));
+		dmu_write(bpo->bpo_os, bpo->bpo_object,
```
[the remark this answers is not in the document]
Comment (ahrens) [reply module/zfs/bpobj.c]: same here

File: module/zfs/bpobj.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (PaulZ-98) [reply module/zfs/bpobj.c]: Got it.  So is it the tests at 513 and 525 that make sure we don't trip the VERIFY at 539 or 570 (i.e. numsubsub and numbps fit in one dbuf)?

[the remark this answers is not in the document]
Comment (ahrens) [reply module/zfs/bpobj.c]: that's right.

File: module/zfs/bpobj.c [hunk 3]
```diff
@@ -377,11 +377,98 @@ bpobj_iterate_nofree(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx)
...
+ * | sub-bpobj    |----------------------> | subsubobj    |  ^
+ * +----+----+----+----+---------+----+    +-----+-----+--+  |
+ * | bp | bp | bp | bp |   ...   | bp |    | OBJ | OBJ |-->-/
+ * +----+----+----+----+---------+----+    +-----+-----+
+ *
+ * Result: sub-sub-bpobjs and subobj added to parent's subobj list.
+ * +--------------+                     +--------------+
+ * | bpobj        |-------------------->| subobj list  |
+ * +----+----+----+----+----+           +-----+-----+--+--+-----+-----+-----+
+ * | bp | bp | bp | bp | bp |           | obj | obj | obj | OBJ | OBJ | OBJ |
```
[the remark this answers is not in the document]
Comment (ahrens) [reply module/zfs/bpobj.c]: Sure.

**Comment by ahrens** [comment]
@sdimitro Verification was mostly with the test suite + ztest, but I did also spot check with the script in https://github.com/zfsonlinux/zfs/issues/7908 and zdb.  The script doesn't reproduce https://github.com/zfsonlinux/zfs/issues/7908 with this fix, because there aren't enough BP's in each bpobj for them to be kept separate now.

**Comment by ahrens** [comment]
@behlendorf rebased.

**Pull request merged**